src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs

```diff
@@ -0,0 +1,546 @@
+            var container = CreateContainer(collection);
+
+            // Act
+            var service1 = container.GetService<IFakeSingletonService>();
+            var service2 = container.GetService<IFakeSingletonService>();
+
+            // Assert
+            Assert.NotNull(service1);
+            Assert.Same(service1, service2);
+        }
+
+        [Fact]
+        public void ServiceContainerRegistersServiceScopeFactory()
+        {
+            // Arrange
+            var collection = new ServiceCollection();
+            var container = CreateContainer(collection);
+
+            // Act
+            var scopeFactory = container.GetService<IServiceScopeFactory>();
+
+            // Assert
+            Assert.NotNull(scopeFactory);
+        }
+
+        [Fact]
+        public void ScopedServiceCanBeResolved()
+        {
+            // Arrange
+            var collection = new ServiceCollection();
+            collection.AddScoped<IFakeScopedService, FakeService>();
+            var container = CreateContainer(collection);
+
+            // Act
+            var scopeFactory = container.GetRequiredService<IServiceScopeFactory>();
+            using (var scope = scopeFactory.CreateScope())
+            {
+                var containerScopedService = container.GetService<IFakeScopedService>();
+                var scopedService1 = scope.ServiceProvider.GetService<IFakeScopedService>();
+                var scopedService2 = scope.ServiceProvider.GetService<IFakeScopedService>();
+
+                // Assert
+                Assert.NotEqual(containerScopedService, scopedService1);
+                Assert.Equal(scopedService1, scopedService2);
+            }
+        }
+
+        [Fact]
+        public void NestedScopedServiceCanBeResolved()
+        {
+            // Arrange
+            var collection = new ServiceCollection();
+            collection.AddScoped<IFakeScopedService, FakeService>();
+            var container = CreateContainer(collection);
+
+            // Act
+            var outerScopeFactory = container.GetService<IServiceScopeFactory>();
+            using (var outerScope = outerScopeFactory.CreateScope())
+            {
+                var innerScopeFactory = outerScope.ServiceProvider.GetService<IServiceScopeFactory>();
+                using (var innerScope = innerScopeFactory.CreateScope())
+                {
+                    var outerScopedService = outerScope.ServiceProvider.GetService<IFakeScopedService>();
+                    var innerScopedService = innerScope.ServiceProvider.GetService<IFakeScopedService>();
+
+                    // Assert
+                    Assert.NotNull(outerScopedService);
+                    Assert.NotNull(innerScopedService);
+                    Assert.NotEqual(outerScopedService, innerScopedService);
+                }
+            }
+        }
+
+        [Fact]
+        public void DisposingScopeDisposesService()
```

Comment:
Would throwing from FakeService if it's already disposed suffice for this scenario? Or am I missing something?